### PR TITLE
Add pluginlib interfaces for using different solvers and graph visualizations

### PR DESCRIPTION
**Pull request by alain-m**
- This PR proposes to add two pluginlib interfaces for slam_karto that facilitate using different solvers than SPA, and different graph visualizations.  If the specified plugins cannot be loaded, fallback behavior is set to load the default plugins (SPA). If it fails too, the node is shutdown.
- Header files are also moved to the include/ folder following http://www.ros.org/reps/rep-0122.html#include
- To test loading the plugins, run the slam_karto node with different `solver_type` and `visualizer_type` parameters.

Credits : @jonbinney , Liz Murphy

### CODE REVIEW

**Comment by alain-m** [comment]
@mikeferguson is there anything preventing from merging this ? Let us know.

**Comment by stwirth** [comment]
@mikeferguson 
giphy

File: include/slam_karto/graph_visualizer.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#ifndef GRAPH_VISUALIZER_H
+#define GRAPH_VISUALIZER_H
+
+#include <open_karto/Mapper.h>
+#include <visualization_msgs/MarkerArray.h>
+#include <string>
+
+namespace karto {
```
Comment (lucbettaieb) [inline include/slam_karto/graph_visualizer.h:8]: I think this needs to be linted.  Please run the ros linter and make sure it's happy, or at least happy enough.  Some of the things it recommends aren't really great.

File: include/slam_karto/graph_visualizer.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  {
+    map_frame_id_ = frame_id;
+  }
+ protected:
+  std::string map_frame_id_;
+};
+
+}; // namespace karto
+
+#endif // GRAPH_VISUALIZER_H
```
Comment (lucbettaieb) [inline include/slam_karto/graph_visualizer.h:29]: Make sure you have newlines at the end of your source files.  It helps render the diffs.

File: src/slam_karto.cpp [hunk 3]
```diff
@@ -36,16 +36,19 @@
 #include "sensor_msgs/LaserScan.h"
```
Comment (lucbettaieb) [inline src/slam_karto.cpp]: May as well make these <'s too?

File: src/slam_karto.cpp [hunk 4]
```diff
@@ -717,6 +700,19 @@ SlamKarto::mapCallback(nav_msgs::GetMap::Request  &req,
     return false;
 }
 
+void
+SlamKarto::visLoop(double vis_publish_period)
+{
+  if (vis_publish_period == 0)
```
Comment (lucbettaieb) [inline src/slam_karto.cpp]: might be clearer to do 0.0 or make it clear 0 is a double.

File: src/spa_graph_visualizer.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  edge.scale.y = 0.1;
+  edge.scale.z = 0.1;
+  edge.color.a = 1.0;
+  edge.color.r = 0.0;
+  edge.color.g = 0.0;
+  edge.color.b = 1.0;
+
+  m.action = visualization_msgs::Marker::ADD;
+  uint id = 0;
+  for (uint i=0; i<graph.size()/2; i++)
```
Comment (lucbettaieb) [inline src/spa_graph_visualizer.cpp]: `size_t` would be good here.

File: src/spa_solver.cpp [hunk 6]
```diff
@@ -15,32 +15,32 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include "spa_solver.h"
+#include <slam_karto/spa_solver.h>
+#include "ros/console.h"
```
Comment (lucbettaieb) [inline src/spa_solver.cpp]: `<`?

**Review by lucbettaieb — CHANGES_REQUESTED** [review]
LGTM for the most part!  Small nitpicks, but we should really try to adhere to the ROS style guide for the most part when possible.

I'll test it on my laptop soon, and hopefully we can get this merged in soon.

**Comment by lucbettaieb** [comment]
@alain-m assign this back to me when you're done.

**Comment by lucbettaieb** [comment]
@alain-m psssssst

**Comment by lucbettaieb** [comment]
@alain-m @jonbinney ?

**Comment by lucbettaieb** [comment]
@alain-m ?

also ping @mikeferguson

**Comment by alain-m** [comment]
@lucbettaieb thanks a lot for the review. Sorry for the delay, I have had a long interruption from work over the past year, I am just catching up. I updated the PR to address your comments. Let me know if you need further improvements.

**Comment by lucbettaieb** [comment]
No problem, glad to hear from you.  I'll take a look hopefully this week.

**Comment by lucbettaieb** [comment]
@alain-m ... one year later...

Do you have a bagfile I could test this with?

**Comment by lucbettaieb** [comment]
This is super stale.  Closing.  Sorry it never worked out ;c

**Comment by alain-m** [comment]
@lucbettaieb sorry for radio silence again, I had some more interruptions at work and don't have time to work on refreshing this at the moment. Thanks for the review anyway.

**Comment by lucbettaieb** [comment]
@alain-m No problem.  If you ever would like to revisit this, please open it back up.

**Pull request closed without merging**